### PR TITLE
fs: TS

### PR DESCRIPTION
**Pull request by jakob223**
- [x] types
- [x] constants
- [x] merge #17692 
- [x] rebase on master 
- [x] common
- [x] simple-screens
- [x] banner
- [x] footer
- [x] nav-header
- [x] send-to-chat
- [x] top-bar
- [x] filepreview
- [x] browser
- [x] top level

### CODE REVIEW

File: shared/fs/top-bar/sort.tsx [hunk 1]
```diff
@@ -1,17 +1,16 @@
...
 
-export type SortBarProps = {|
-  sortByNameAsc?: ?() => void,
-  sortByNameDesc?: ?() => void,
-  sortByTimeAsc?: ?() => void,
-  sortByTimeDesc?: ?() => void,
-  sortSetting?: ?Types.SortSetting,
-|}
+export type SortBarProps = {
+  sortByNameAsc?: () => void | null
```
Comment (jakob223) [inline shared/fs/top-bar/sort.tsx]: Does precedence work as expected here? I have a suspicion that TS reads this as `() => (void | null)`

File: shared/fs/nav-header/index.stories.tsx [hunk 2]
```diff
@@ -11,10 +10,12 @@ import FilesContainer from '../container'
 
 export const headerProvider = {
   MainBanner: (p: any) => ({
-    bannerType: p.storyProps?.bannerType || 'none',
+    // Auto generated from flowToTs. Please clean me!
+    bannerType:
+      (p.storyProps === null || p.storyProps === undefined ? undefined : p.storyProps.bannerType) || 'none',
```
Comment (jakob223) [inline shared/fs/nav-header/index.stories.tsx]: I wrote this code, will deal with this.

File: shared/fs/nav-header/mobile-header.tsx [hunk 3]
```diff
@@ -8,13 +7,14 @@ import Actions from './actions'
...
 
-type Props = {|
-  onBack: ?() => void,
-  path: Types.Path,
-|}
-type State = {|
-  filterExpanded: boolean,
-|}
+type Props = {
+  onBack: () => void | null
```
Comment (jakob223) [inline shared/fs/nav-header/mobile-header.tsx]: same as above

File: shared/fs/send-to-chat/attachment/choose-conversation-hoc.tsx [hunk 4]
```diff
@@ -24,24 +23,14 @@ const mergeProps = (stateProps, dispatchProps, ownProps) => ({
...
+  onSelect: (convID: ChatTypes.ConversationIDKey) => void
+  onSetFilter?: (filter: string) => void
+  selected: ChatTypes.ConversationIDKey
 }
 
-type WithInjectedProps<OriginalProps> = {|
-  ...$Exact<OriginalProps>,
-  ...$Exact<InjectedProps>,
-|}
+type WithInjectedProps<OriginalProps> = {} & OriginalProps & InjectedProps
```
Comment (jakob223) [inline shared/fs/send-to-chat/attachment/choose-conversation-hoc.tsx]: Does there need to be a `{}` here?

File: shared/fs/footer/downloads-container.tsx [hunk 5]
```diff
@@ -24,15 +23,10 @@ const mergeProps = ({_downloads}, {openDownloadFolder}) => {
       return b.state.startedAt - a.state.startedAt // newer first
     })
     .map(([key, download]) => key)
-  return ({
+  return {
     downloadKeys,
     openDownloadFolder,
-  }: DownloadsProps)
+  } as DownloadsProps
```
Comment (jakob223) [inline shared/fs/footer/downloads-container.tsx]: Why do we need this `as`? Can you give the return type of `mergeProps` in its declaration instead?

I admit I don't fully understand TS's type inference but I found it useful in once place to fill in the template parameters to `namedConnect`, maybe that would help avoid this?

File: shared/fs/footer/upload-countdown-hoc.tsx [hunk 6]
```diff
@@ -1,71 +1,72 @@
...
 type Props = PropsWithTimer<UploadCountdownHOCProps>
 
 // Cosider this component as a state machine with following four states. 1Hz
 // Ticks (from _tick() calls by setInterval) and props changes (through
 // componentDidUpdate() calls) are two possible inputs.
-type Mode =
+enum Mode {
   // The upload banner isn't shown.
-  | 'hidden'
+  Hidden = 1,
```
Comment (jakob223) [inline shared/fs/footer/upload-countdown-hoc.tsx]: why `= 1`?

File: shared/fs/banner/reset-banner/container.tsx [hunk 7]
```diff
@@ -11,15 +10,17 @@ import {folderNameWithoutUsers} from '../../../util/kbfs'
...
+}
 
 const mapStateToProps = (state, {path}: OwnProps) => ({
   _tlf: Constants.getTlfFromPath(state.fs.tlfs, path),
   _username: state.config.username,
 })
 
 const mapDispatchToProps = dispatch => ({
-  _onOpenWithoutResetUsers: (currPath: Types.Path, users: {[string]: boolean}) => {
+  _onOpenWithoutResetUsers: (currPath: Types.Path, users: {[K in string]: boolean}) => {
```
Comment (jakob223) [inline shared/fs/banner/reset-banner/container.tsx]: what's the `K in` doing?

File: shared/fs/banner/system-file-manager-integration-banner/kext-permission-popup.tsx [hunk 8]
```diff
@@ -1,14 +1,13 @@
...
 import * as Styles from '../../../styles'
 
-type Props = {|
-  driverStatus: Types.DriverStatus,
-  onCancel: () => void, // must be ownProps.navigateUp() or idempotent in a way
-  openSecurityPrefs: () => void,
-|}
+type Props = {
+  driverStatus: Types.DriverStatus
+  onCancel: () => void // must be ownProps.navigateUp() or idempotent in a way,
```
Comment (jakob223) [inline shared/fs/banner/system-file-manager-integration-banner/kext-permission-popup.tsx]: why the added comma? is this a flowToTs bug?

File: shared/constants/fs.tsx [hunk 9]
```diff
@@ -202,15 +202,15 @@ const _makeError: I.Record.Factory<Types._FsError> = I.Record({
   time: 0,
 })
 
-type Record = {
+type _MakeErrorArgs = {
```
Comment (jakob223) [inline shared/constants/fs.tsx]: 👍

File: shared/fs/top-bar/sort.tsx (hunk 1 above)
Comment (songgao) [reply shared/fs/top-bar/sort.tsx]: good catch!

File: shared/fs/nav-header/index.stories.tsx (hunk 2 above)
Comment (songgao) [reply shared/fs/nav-header/index.stories.tsx]: oops I missed that flowToTs comment

File: shared/fs/footer/downloads-container.tsx (hunk 5 above)
Comment (songgao) [reply shared/fs/footer/downloads-container.tsx]: I think this is just a direct translation from the old code. I'll update it to have it in the declaration.

File: shared/fs/footer/upload-countdown-hoc.tsx (hunk 6 above)
Comment (songgao) [reply shared/fs/footer/upload-countdown-hoc.tsx]: Not necessary; removing. I was (incorrectly) under the impression this was needed for it to be integer underneath. Just realized it just defaults to start from 0 if not assigned a thing.

File: shared/fs/banner/reset-banner/container.tsx (hunk 7 above)
Comment (songgao) [reply shared/fs/banner/reset-banner/container.tsx]: Also from flowToTs. Although I just tried; without `K in` it wouldn't work. I think this is [Mapped types](https://www.typescriptlang.org/docs/handbook/advanced-types.html#mapped-types): https://www.typescriptlang.org/docs/handbook/advanced-types.html#mapped-types

File: shared/fs/banner/system-file-manager-integration-banner/kext-permission-popup.tsx (hunk 8 above)
Comment (songgao) [reply shared/fs/banner/system-file-manager-integration-banner/kext-permission-popup.tsx]: hmm not sure; possibly. Will get rid of it.

File: shared/fs/footer/downloads-container.tsx [hunk 10]
```diff
@@ -1,8 +1,10 @@
 import {namedConnect} from '../../util/container'
 import * as FsGen from '../../actions/fs-gen'
+import * as Types from '../../constants/types/fs'
 import Downloads, {DownloadsProps} from './downloads'
 import {isMobile} from '../../constants/platform'
 import {downloadFolder} from '../../util/file'
+import {types} from '@babel/core'
```
Comment (jakob223) [inline shared/fs/footer/downloads-container.tsx]: Is this used?

Comment (songgao) [reply shared/fs/footer/downloads-container.tsx]: good catch! not sure how it got in there or why the linter didn't catch it

**Pull request merged**